website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+    while True:
+        question = input("\nEnter your SQL question (or type 'exit' to quit): ")
+        if question.lower() == "exit":
+            break
+
+        # Initialize the state with all required keys
+        initial_state = {
+            "messages": [("user", question)],
+            "iterations": 0,
+            "error": "",
+            "results": None,
+            "generation": None,
+            "no_records_found": False,
+            "translated_input": "",  # Initialize translated_input
+        }
+
+        solution = app.invoke(initial_state)
+
+        # Check if an error was set during the safety check
+        if solution["error"] == "yes":
+            print("\nAssistant Message:\n")
+            print(solution["messages"][-1][1])  # Display the assistant's message
+            continue  # Skip to the next iteration
+
+        # Extract the generated SQL query from solution["generation"]
+        sql_query = solution["generation"].sql_code
+        print("\nGenerated SQL Query:\n")
+        print(sql_query)
+
+        # Extract and display the query results
+        if solution.get("no_records_found"):
+            print("\nNo records found matching your query.")
+        elif "results" in solution and solution["results"] is not None:
+            print("\nQuery Results:\n")
+            for row in solution["results"]:
+                print(row)
+        else:
+            print("\nNo results returned or query did not execute successfully.")
+
+    print("Goodbye!")
+
+
+if __name__ == "__main__":
+    main()
+
+```
+
+# Project File Structure
+
+The Project follows a simple file structure:
+
+<img src="file_structure.png" alt="file_structure" width="300">
+
+# MLFlow Tracing
+
+[MLflow Automated Tracing](https://mlflow.org/docs/latest/llms/tracing/index.html) provides fully automated integrations with various GenAI libraries such as LangChain, OpenAI, LlamaIndex, DSPy, and AutoGen. Since our AI Workflow is built using LangGraph, we can activate automated LangChain tracing by enabling `mlflow.langchain.autolog()`.
+
+With LangChain autologging, traces are automatically logged to the active MLflow experiment whenever invocation APIs are called on chains. This seamless integration ensures that every interaction is captured for monitoring and analysis.
+
+## Viewing Traces in MLflow
+
+Traces can be easily accessed by navigating to the MLflow experiment of interest and clicking on the "Tracing" tab. Once inside, selecting a specific trace provides detailed execution information.
+
+Each trace includes:
+
+1. **Execution Graphs**: Visualizations of the workflow steps.
+2. **Inputs and Outputs**: Detailed logs of data processed at each step.
+
+This granular visibility enables developers to debug and optimize their workflows effectively.
+
+By leveraging MLflow tracing, we ensure that our Multi-Lingual Query Engine remains transparent, auditable, and scalable.
+
+![mlflow_tracing_gif](mlflow_trace.gif)
+
```

Comment:
I really like this, would love some more emphasis on how tracing lets you visualize the whole graph execution.

Reply:
I’ve added more emphasis on how MLflow Tracing provides a comprehensive view of the entire graph execution.